frame/executive/src/lib.rs

```diff
@@ -349,12 +357,39 @@ where
 	/// Execute given extrinsics and take care of post-extrinsics book-keeping.
 	fn execute_extrinsics_with_book_keeping(
 		extrinsics: Vec<Block::Extrinsic>,
+		maybe_first_signed_index: Option<u32>,
 		block_number: NumberFor<Block>,
 	) {
-		extrinsics.into_iter().for_each(|e| if let Err(e) = Self::apply_extrinsic(e) {
-			let err: &'static str = e.into();
-			panic!("{}", err)
-		});
+		let do_on_post_inherent = || {
+			let mut weight = 0;
+			weight = weight.saturating_add(
+				<frame_system::Pallet<System> as OnPostInherent<System::BlockNumber>>::on_post_inherent(block_number)
+			);
+			weight = weight.saturating_add(
+				<AllPallets as OnPostInherent<System::BlockNumber>>::on_post_inherent(block_number)
+			);
+			<frame_system::Pallet::<System>>::register_extra_weight_unchecked(weight, DispatchClass::Mandatory);
```

Comment:
it doesn't seem very correct to me to have weight on post_inherent hook:
the inherent are not all mandatory, if a block producer include a non-mandatory inherent which makes the block full, and then if a post_inherent hook use even more weight, the non-mandatory inherent shouldn't have been in the block in the first place.

I think similarly to on_finalize, the on_post_inherent should return no weight, no ?

Reply:
what about the scenario where on_post_inherent fills up the blocks remaining weight; if we don't count that weight then we might start adding extrinsics and then even on_finalize when the block is already full - seems problematic?

Is there a way to include the weight from on_post_inherent but make sure it executes regardless if it pushes the block to become overweight (essentially the equivalent of Operational dispatch)?

Reply:
First i am making the assumption that having non-mandatory inherent is allowed, if all inherent must be mandatory then my comment is wrong and on_post_inherent can return some mandatory weight.

> what about the scenario where on_post_inherent fills up the blocks remaining weight; if we don't count that weight then we might start adding extrinsics and then even on_finalize when the block is already full - seems problematic?

Yes the on_post_inherent weight must be taken into account, and returned by on_initialize, similarly to on_finalize which doesn't return its weight, but whose weight must be returned by on_initialize.

The scenario where on_post_inherent fills up the blocks is exaclty the issue here: if there is some non-mandatory inherents, block producer can include them regardless of the weight of on_post_inherent. But what we want is that if on_post_inherent fills up the block then no non-mandatory inherents are in the block.

> Is there a way to include the weight from on_post_inherent but make sure it executes regardless if it pushes the block to become overweight (essentially the equivalent of Operational dispatch)?

Yes this is Mandatory dispatch class.